docs(tree): staged allowed types usage guide (#25581)
microsoft/FluidFramework: docs/docs/data-structures/tree/schema-evolution/allowed-types-rollout.mdx

---
title: Rolling Out New Allowed Types
sidebar_position: 3
---

# Rolling Out New Allowed Types

> This guide shows you how to safely add new allowed types to existing fields in your [SharedTree](../index.mdx) schema without breaking [forwards compatibility](./index.mdx#forwards-compatibility) with existing applications.

Reviewer: There are a couple issues with this:
1. I think it would be good to link the relevant API docs, for example https://fluidframework.com/docs/api/tree/allowedtypes-typealias
2. The use of "fields" here is questionable since this process should also handle maps and array's.
3. I think we should avoid second person in docs for a more formal tone and also to imply one can use this to modify other people's schema as well.

Maybe:

```suggestion
This guide shows how to safely add a new [TreeNodeSchema](https://fluidframework.com/docs/api/fluid-framework/treenodeschema-typealias) to an existing [AllowedTypes](https://fluidframework.com/docs/api/tree/allowedtypes-typealias) in a [SharedTree](../index.mdx) schema without breaking [forwards compatibility](./index.mdx#forwards-compatibility) with existing applications.
```

We can now deploy `schemaB` to clients as a complete replacement to `schemaA`.
At this point, new clients will be able to read documents containing strings and old documents will not.
This is safe because no clients have the ability to insert strings into the document.

```typescript
// view second tree with schema B
const configB = new TreeViewConfiguration({
    schema: schemaB,
});
const viewB = treeB.viewWith(configB);
// check that we can read the tree
assert.deepEqual(viewB.root, 5);
// upgrade to schema B: this is a no-op
viewB.upgradeSchema();
synchronizeTrees();

// check view A can read the document
assert.deepEqual(viewA.root, 5);
// check view B cannot write strings to the root
assert.throws(() => {
    viewB.root = "test";
});
```

For this schema change, the schema does not need to be [upgraded](./index.mdx#schema-upgrade-process) because [`staged`](../../../api/fluid-framework/schemafactoryalpha-class#staged-property) allowed types are removed from stored schemas.
Upgrading the schema would result in a noop.

### Step 3: Wait for Client Saturation

Ensure all clients in your deployment have the staged support before proceeding.

> Monitor your deployment to confirm coverage.

Reviewer: ```suggestion
Monitor the deployment to confirm coverage.
```
